working-w-copilot/copilot-in-projects-refactoring.md

```diff
@@ -0,0 +1,612 @@
+- `test_routes.py`
+
+Start in `routes.py` and use the cursor to highlight the entire `validate_book` function. Next, bring up the inline Copilot chat and invoke the `/tests` shortcut. When Copilot is done thinking, we may see a slightly different UI than in previous scenarios:
+
+![A VS Code window with the routes.py file up and validate_book function selected showing a Refactor Preview tab at the bottom of the screen](assets/copilot-in-projects/routes-test-validate-book.png)  
+*Fig. Copilot's `Refactor Preview` tab showing at the bottom of VS Code*
+
+This time, instead of showing changes to the current file in a temporary window, the bottom panel has opened to a `Refactor Preview` tab that has a single entry on it. If we click the entry, it will take us to a temporary file showing the changes Copilot is suggesting that we make to the `test_routes.py` file.
+
+![A VS Code window with a temporary file showing that contains the original contents of test_routes.py along with suggested changes to test the validate_book function](assets/copilot-in-projects/test-routes-validate-book-refactor-preview.png)  
+*Fig. Temporary file in VS Code showing the test suggestions from Copilot*
+
+Here we can review the test cases Copilot generated and see if there are changes or further tests we need. We have "Apply" and "Discard" controls at the bottom of the screen in the `Refactor Preview` tab to either accept or reject the changes.
+
+Examining the code created by Copilot, the scenarios identified are great; we have tests for the nominal case and a couple edge cases of invalid or non-existent book ids. However, there are a number of issues:
+- imports were added in the middle of the file, one of which is tacked on to the last line of an existing test
+- `test_validate_book_with_valid_id` doesn't use our `two_saved_books` test fixture to ensure that there are books in the database before the test runs
+- `test_validate_book_with_valid_id` thinks the result of `validate_book(book_id)` will be an integer rather than a `Book` instance
+- `test_validate_book_with_invalid_id` and `test_validate_book_with_nonexistent_id` try to read the error value from outside the `with pytest.raises` block where it's out of scope
+
+### !callout-info
+
+## Be Sure to Review Your Own Output
+
+Remember, the tests generated during the Copilot run in this lesson may not match what you see when trying on your own. Always closely review any suggestions from Copilot or any other AI tool to ensure they meet your needs and are correct for your project.
+
+### !end-callout
+
+Since we want these test scenarios and the tests bodies are pretty close to what we're looking for, let's use the "Apply" button in the `Refactor Preview` tab to accept the changes. We'll address the issues listed above by adjusting the imports, adding the relevant test fixtures, and updating the assertions to ensure they are for the correct values and are located in scope of the objects they are trying to examine.
```

Comment:
- fully comma-separate the subclause
```suggestion
Since we want these test scenarios, and the tests bodies are pretty close to what we're looking for, let's use the "Apply" button in the `Refactor Preview` tab to accept the changes. We'll address the issues listed above by adjusting the imports, adding the relevant test fixtures, and updating the assertions to ensure they are for the correct values and are located in scope of the objects they are trying to examine.
```